pkg/rpcserver/rpcserver.go

```diff
@@ -306,25 +328,23 @@ func (serv *server) handleRunnerConn(ctx context.Context, eg *errgroup.Group,
 		return err
 	}
 
-	serv.checkOnce.Do(func() {
-		// Run the machine check.
-		eg.Go(func() error {
-			if err := serv.runCheck(ctx, &info); err != nil {
-				return fmt.Errorf("%w: %w", errFatal, err)
-			}
-			return nil
-		})
-	})
-
-	if serv.triagedCorpus.Load() {
-		eg.Go(runner.SendCorpusTriaged)
+	select {
+	case <-ctx.Done():
+	case onHandshake <- &info:
 	}
 
+	ctx2, cancel := context.WithCancel(ctx)
+	eg, ctx3 := errgroup.WithContext(ctx2)
+	eg.Go(runner.SendCorpusTriaged)
+	eg.Go(func() error {
+		defer cancel()
+		return serv.connectionLoop(runner)
+	})
 	go func() {
-		<-ctx.Done()
+		<-ctx3.Done()
```

Comment:
When is ctx3.Done() ready? Does it return when when connectionLoop has returned?

I can't say I understand much here now with this mix of multiple contexts, errgroups, cancels, waits, and goroutines. This looks very complex and fragile.